src/convert.rs

```diff
@@ -0,0 +1,200 @@
+use std::convert::TryInto;
+
+use crate::color::{ExtendedColorType, ExtendedColorType::*};
+
+/// Error from trying to convert a buffer of pixels.
+#[derive(Copy, Clone, Debug)]
+pub enum ConvertError {
+    /// In input pixels are invalid for the given color type.
+    CorruptInput,
+    /// The alignment of the provided buffers doesn't match the required alignment for the
+    /// converter.
+    BadAlignment,
+    /// The parameters specifying width and row pitch are inconsistent.
+    BadDimensions,
+}
+
+/// An object that can convert raw buffers of pixels from one format to another.
+pub trait Converter {
+    /// Returns the (source, destination) color types that this converts between.
+    fn color_types(&self) -> (ExtendedColorType, ExtendedColorType);
+
+    /// Returns the require alignments for source and destination buffers.
+    fn required_alignments(&self) -> (u8, u8);
+
+    /// Perform the conversion.
+    ///
+    /// * `source` contains the input pixels and must have alignment of at least
+    ///   `require_alignment().0`.
+    ///
+    /// * `source_width` is the number of pixels in a source row (the number of rows is computed
+    ///   based on the length of the buffer).
+    ///
+    /// * `source_row_pitch` is the number of bytes between the start of successive rows in the
+    ///   source buffer, or None if the pixels are tighly packed. Must be divisible by the source
+    ///   buffer's required alignment.
+    ///
+    /// * `destination` is the output buffer to use which must be aligned to at least
+    ///   `required_alignment().1`.
+    ///
+    /// * `destination_row_pitch` is the number of bytes between the start of successive rows in the
+    ///   destination buffer, or None if the pixels are tightly packed. Must be divisible by the
+    ///   destination buffer's required alignment.
+    fn convert(
+        &self,
+        source: &[u8],
+        source_width: u32,
+        source_row_pitch: usize,
+        destination: &mut [u8],
+        destination_row_pitch: usize,
+    ) -> Result<(), ConvertError>;
```

Comment:
The thinking here is that there is only a single allowed layout: image rows start on a byte boundary and pixels are packed with no space between them. The reason to include _source_row_pitch_ and _destination_row_pitch_ is to allow operating on sub-rectangles within an image.

To make this usable we probably should make ExtendedColorType::bits_per_pixel public, but I'm not sure whether it makes sense to add more functions to compute byte sizes. The key difference with ImageDecoder is that the image data (and by extension the dimensions) are being provided by the user.

Reply:
I see. That makes a lot of sense. Another way would be to provide a small layout descriptor struct to either wrap or accompany the slices. That would hopefully also make taking the step to have a pixel pitch cheaper to use (admittedly being more complicated to implement though). Actually, what if we used the existing [`SampleLayout`](https://docs.rs/image/0.23.6/image/flat/struct.SampleLayout.html) or should we check it for consistency before passing it as argument?

Reply:
I like the idea of using `SampleLayout` though we should think through how that interacts with formats that have channels with less than one byte per pixel. Perhaps by making all the strides actually being in terms of bits-per-channel?

It also might make sense to allow converters to reject formats they don't like, so we don't get bogged down supporting weird layouts. That could either happen via a constructor, or just by returning an error from convert. In either case the documentation should clearly state what layouts are and aren't allowed.

Reply:
So `SampleLayout` is indeed a subset of the possible layouts we want. I think we should introduce an opaque wrapper until we are more confident what we really need then, that can be converted `From<SampleLayout>`. It would allow us to choose its own interface as well. We have to be careful not to be too opaque but it seems not too hard to _at least_ provide the information passed as parameters in the above implementation.